### PR TITLE
Internal: Support light DOM in atomic elements [EDS-615]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

**Comment by louiswol94** [comment]
This is so CLEAN! :)

File: modules/atomic-widgets/assets/js/editor/module.js [hunk 1]
```diff
@@ -1,8 +1,19 @@
...
 		$e.components.register( new Component() );
+
+		this.registerAtomicWidgetTypes();
+	}
+
+	registerAtomicWidgetTypes() {
+		Object.entries( elementor.widgetsCache ?? {} )
+			.filter( ( [ , widget ] ) => !! widget.atomic )
+			.forEach(
+				( [ type ] ) => elementor.elementsManager.registerElementType( new AtomicWidgetType( type ) ),
```
Comment (louiswol94) [inline modules/atomic-widgets/assets/js/editor/module.js:15]: @Nevoss Does every single widget have to register it's own "element type"? 🤔

Comment (Nevoss) [reply modules/atomic-widgets/assets/js/editor/module.js:15]: We can do it, but we have to change the `core` functionality of the `elementsManager` and I did not want to change much in the `core`.

Comment (Nevoss) [reply modules/atomic-widgets/assets/js/editor/module.js:15]: Yeah, did not find another way.

Comment (Nevoss) [reply modules/atomic-widgets/assets/js/editor/module.js:15]: Yes it should works, but I am not sure it is safe to treat the `atomic-widget` as a new element type. but maybe it is fine.

Comment (louiswol94) [reply modules/atomic-widgets/assets/js/editor/module.js:15]: @Nevoss Can you please try it for now and see if it works fine? I feel better about one new element type than multiple.

File: modules/atomic-widgets/assets/js/editor/atomic-widget-type.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { AtomicWidgetView } from './atomic-widget-view';
+
+export class AtomicWidgetType extends elementor.modules.elements.types.Widget {
+	_type;
```
Comment (StyleShit) [inline modules/atomic-widgets/assets/js/editor/atomic-widget-type.js:4]:
```suggestion
	#type;
```

File: modules/atomic-widgets/base/atomic-widget-base.php [hunk 3]
```diff
@@ -173,6 +174,10 @@ private function sanitize_atomic_settings( array $settings ): array {
 		return $validated;
 	}
 
+	public function before_render() {}
+
+	public function after_render() {}
```
Comment (StyleShit) [inline modules/atomic-widgets/base/atomic-widget-base.php]: add comments

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+export class AtomicWidgetView extends elementor.modules.elements.views.Widget {
+	// Dispatch `render` event for so the overlay layer can be updated
+	onRender( ...args ) {
+		super.onRender( ...args );
+
+		this._dispatchEvent( 'elementor/preview/atomic-widget/render' );
```
Comment (StyleShit) [inline modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:6]:
```suggestion
		this.#dispatchEvent( 'elementor/preview/atomic-widget/render' );
```

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
+export class AtomicWidgetView extends elementor.modules.elements.views.Widget {
+	// Dispatch `render` event for so the overlay layer can be updated
```
Comment (StyleShit) [inline modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:2]:
```suggestion
	// Dispatch `render` event so the overlay layer will be updated
```

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	// Dispatch `destroy` event for so the overlay layer can be updated
+	onDestroy( ...args ) {
+		super.onDestroy( ...args );
+
+		this._dispatchEvent( 'elementor/preview/atomic-widget/destroy' );
+	}
+
+	// Removes behaviors that are not needed for atomic widgets (Should be reimplemented in the overlay layer).
```
Comment (StyleShit) [inline modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:16]:
```suggestion
	// Removes behaviors that are not needed for atomic widgets (that are implemented in the overlay layer).
```

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	getHandlesOverlay() {
+		return null;
+	}
+
+	// Set the `data-atomic` attribute to the widget element + `display: contents;`.
+	// The css will behave like the wrapper `div` is not exists and with that we can remove this `div` in the frontend.
+	attributes() {
+		return {
+			...super.attributes(),
+			style: 'display: contents !important;',
```
Comment (StyleShit) [inline modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:41]:
```suggestion
			style: 'display: contents !important;', // Make the wrapper non-existent in terms of CSS to mimic the frontend DOM tree.
```

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+		return this.$el.find( ':first-child' );
+	}
+
+	// Remove the overlay, so we can use the new overlay layer.
+	getHandlesOverlay() {
+		return null;
+	}
+
+	// Set the `data-atomic` attribute to the widget element + `display: contents;`.
+	// The css will behave like the wrapper `div` is not exists and with that we can remove this `div` in the frontend.
```
Comment (StyleShit) [inline modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:37]:
```suggestion
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: modules/atomic-widgets/base/atomic-widget-base.php (hunk 3 above)
Comment (Nevoss) [reply modules/atomic-widgets/base/atomic-widget-base.php]: Done

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js (hunk 4 above)
Comment (Nevoss) [reply modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:6]: Done

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js (hunk 5 above)
Comment (Nevoss) [reply modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:2]: Done

File: modules/atomic-widgets/assets/js/editor/atomic-widget-view.js (hunk 8 above)
Comment (Nevoss) [reply modules/atomic-widgets/assets/js/editor/atomic-widget-view.js:37]: Done

**Review by StyleShit — APPROVED** [review]

**Pull request merged**